services/audit-service/src/main/java/datawave/microservice/audit/auditors/file/FileAuditor.java

```diff
@@ -0,0 +1,212 @@
+ */
+public class FileAuditor implements Auditor {
+    
+    private final Logger log = LoggerFactory.getLogger(this.getClass());
+    
+    protected static final ObjectMapper mapper = new ObjectMapper();
+    
+    protected final SimpleDateFormat sdf;
+    protected final ReentrantLock writeLock = new ReentrantLock(true);
+    
+    protected long maxFileLenBytes;
+    protected long maxFileAgeMillis;
+    
+    protected FileSystem fileSystem;
+    protected Path path;
+    
+    protected Path currentFile = null;
+    protected Date creationDate = null;
+    
+    protected FileAuditor(Builder<?> builder) throws URISyntaxException, IOException {
+        this.maxFileLenBytes = builder.maxFileLenBytes;
+        this.maxFileAgeMillis = builder.maxFileAgeMillis;
+        
+        Configuration config = new Configuration();
+        
+        if (builder.configResources != null) {
+            for (String resource : builder.configResources)
+                config.addResource(new Path(resource));
+        }
+        
+        path = new Path(builder.path);
+        
+        if (builder.subpath != null)
+            path = new Path(path, builder.subpath);
+        
+        if (builder.prefix != null)
+            path = new Path(path, builder.prefix);
+        
+        fileSystem = FileSystem.get(path.toUri(), config);
+        
+        if (!fileSystem.exists(path))
+            fileSystem.mkdirs(path);
+        
+        this.sdf = new SimpleDateFormat("yyyyMMdd_HHmmss.SSS'.json'");
+    }
+    
+    @Override
+    public void audit(AuditParameters auditParameters) throws Exception {
+        
+        // convert the messages to URL-encoded JSON
+        Map<String,String> auditParamsMap = auditParameters.toMap();
+        auditParamsMap.forEach((key, value) -> auditParamsMap.put(key, urlEncodeString(value)));
+        String jsonAuditParams = mapper.writeValueAsString(auditParamsMap) + "\n";
+        
+        writeLock.lock();
+        try {
+            // if the file/stream is null, doesn't exist, or the file is too old/big, create a new file & output stream
+            if (currentFile == null || !fileSystem.exists(currentFile) || isFileTooOld() || isFileTooBig())
+                createNewFile();
+            
+            writeAudit(jsonAuditParams);
+        } finally {
+            writeLock.unlock();
+        }
+    }
+    
+    protected String urlEncodeString(String value) {
+        try {
+            return URLEncoder.encode(value, "UTF8");
+        } catch (UnsupportedEncodingException e) {
+            log.error("Unable to URL encode value: {}", value);
+        }
+        return value;
+    }
+    
+    protected void writeAudit(String jsonAuditParams) throws Exception {
+        OutputStream appendStream = (fileSystem instanceof LocalFileSystem) ? new FileOutputStream(new File(currentFile.toUri()), true)
+                        : fileSystem.append(currentFile);
```

Comment:
Just curious... Why the complexity here on testing the file system instance? Does LocalFileSystem not support the append method?

Reply:
It does not, unfortunately.  :(

Reply:
Hmm, weird. It implements the method. Oh well...

Reply:
During testing, it ultimately boiled down to this:
https://github.com/apache/hadoop/blob/release-2.6.0/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/ChecksumFileSystem.java#L352